### DISCUSSION TITLE
Question related to "Physically Informed ML for hydrological modeling under climate nonstationarity"

### DISCUSSION THREAD

**Discussion by Meelisha** [neuralhydrology/neuralhydrology#102] [Q&A]
Hello all!

I had a question specific to this paper [Physically Informed Machine Learning for Hydrological Modeling Under Climate Nonstationarity](https://mdsoar.org/handle/11603/19521) published in 2019. 

Question about the choice of day length to compute climate indices:
The approach defined under "explainable ML" takes in three types of inputs: daily meteorological forcings, static catchment attributes and dynamic climate statistics related to annual precipitation, potential evaporation, etc. The paper explains that the dynamic climate indices were computed on the 365 days just previous to each daily streamflow prediction. Aren't the climate indices informing about the climate of the basins, which I would think is computed with longer time period? What is the significance of choosing 365 days compared to other values (say, 20 years as in [CAMELS climate indices](https://hess.copernicus.org/articles/21/5293/2017/) ) in computing the static attributes? Would you say that the model is somehow learning the trend in the attributes and mapping these to daily runoff prediction? However, if that is the case then, how can we know efficiently that the number of days chosen to calculate the indices is optimum? 

Question about the model and configuration:
How can we define the embedding network that takes static catchment attributes and dynamic climate indices as input, which is then sent to LSTM as input along with the daily met. forcings. Is there a configuration argument in the neuralhydrology package which allows this? 
Are the arguments "additional_feature_files" and "evolving_attributes" enough to replicate the "explainable ML under Climate Nonstationarity" model?
I am not exactly sure how "evolving_attributes" use the data as compared to "dynamic_attributes". What would you recommend to use for assessing the impact of varying climates on streamflow?

**Comment by kratzert** [comment]
Hi @Meelisha,

sorry for the very late reply. However, this forum is meant to be used for discussing things around the NeuralHydrology Python library, not so much for discussing papers, even if by coincidence some of the maintainers here are co-authors. In general, it would be better to simply reach out to the first author, Grey, by email. He will for sure not look at this discussion forum.

I don't have much time, so excuse my short answers:

> The paper explains that the dynamic climate indices were computed on the 365 days just previous to each daily streamflow prediction. Aren't the climate indices informing about the climate of the basins, which I would think is computed with longer time period? What is the significance of choosing 365 days compared to other values (say, 20 years as in CAMELS climate indices ) in computing the static attributes?

That is the entire point of this paper: The idea is to let the model know if a year is particularly dry/wet so that it can maybe, in this particular year, model a otherwise humid basin more as a kind of arid basin (if e.g. a humid basin has two dry years). Of course, the missing wetness can also be inferred from the meteorological inputs, but then the model would only know about this after it has seen large parts of the time series already and not at the beginning of the time series.

> Would you say that the model is somehow learning the trend in the attributes and mapping these to daily runoff prediction? However, if that is the case then, how can we know efficiently that the number of days chosen to calculate the indices is optimum?

That was the idea, yes and you will only know what is the most efficient by doing hyper param tuning.

> How can we define the embedding network that takes static catchment attributes and dynamic climate indices as input, which is then sent to LSTM as input along with the daily met. forcings. Is there a configuration argument in the neuralhydrology package which allows this?

I don't understand this question. If you use any form of static attributes, this is done by default (i.e. that static attributes or the output of the embedding network are added to the LSTM inputs).

> Are the arguments "additional_feature_files" and "evolving_attributes" enough to replicate the "explainable ML under Climate Nonstationarity" model?

Yes.

> I am not exactly sure how "evolving_attributes" use the data as compared to "dynamic_attributes". What would you recommend to use for assessing the impact of varying climates on streamflow?

If you read the definition/documentation of "evolving_attributes", you will understand that they are different to dynamic inputs. Evolving attributes are also stored in a time-indexed dataframe, but for each input sample, one row of these values are taken (by definition the row that aligns with the label) and so you have one static value for each sample (but it can be different for each sample). For dynamic inputs, you will slice an entire time slice for one sample (a number of rows that corresponds to your sequence length).
